Initial commit: Add Universal Document Loader and unit tests

aiml-learning-design opened the pull request:
## Initial Contribution: Universal Document Loader

### What's Added

- `universal_loader.py` inside `universal_document_loader`
- Core class `UniversalDocumentLoader` that selects appropriate loader strategy based on file type.
- Clean, modular design to support multiple formats (PDF, DOCX, RTF, HTML, EML, etc.)

### What's Tested

- Added `test_universal_document_loader.py` under `unit_tests`
- Used `pytest` with parameterized tests for:
  - PDF
  - Word (DOCX)
  - RTF
  - HTML
  - Email (EML, MSG)
  - Markdown
  - PPTX
  - TXT
  - CSV
  - Unknown file formats

- Mocks `os.path.isfile` and individual loader classes to verify correct loader invocation.

### Notes

- First-time contributor 
- Code is modular and follows existing langchain-community patterns.
- Happy to iterate based on feedback 🙏

---

> Thanks for reviewing! Looking forward to being part of the community

eyurtsev reviewed (changes requested):
Hi @aiml-learning-design thank you for the contribution.

I don't want an implementation which hard-codes strategies because there are so many implementations available and I don't know which ones are best for a particular file type.

Switching should probably be based on mimetype by default rather t than extension:

https://github.com/langchain-ai/langchain-community/blob/fdd86e3b8c49c99e5b28e46af204df5d1893197a/libs/community/langchain_community/document_loaders/parsers/generic.py#L14-L14 handles this at the blob layer based on mime-types rather than file-extensions


Why is the RTF file so big? Can you make a minimal representation?

aiml-learning-design commented:
> > Hi @aiml-learning-design thank you for the contribution.
> > I don't want an implementation which hard-codes strategies because there are so many implementations available and I don't know which ones are best for a particular file type.
> > Switching should probably be based on mimetype by default rather t than extension:
> > https://github.com/langchain-ai/langchain-community/blob/fdd86e3b8c49c99e5b28e46af204df5d1893197a/libs/community/langchain_community/document_loaders/parsers/generic.py#L14-L14
> > 
> > handles this at the blob layer based on mime-types rather than file-extensions
> > Why is the RTF file so big? Can you make a minimal representation?
> 
> Thank you for the valuable feedback 🙏
> 
> I agree, hardcoding extension based strategies has limitations, especially as more document types are supported. I’ll refactor the code to use MIME type detection using Python’s mimetypes module (as suggested).
> 
> Regarding the large .rtf file, thanks for pointing out. I’ll replace it with a minimal representation. I’ll push the updates shortly. Thank you for your guidance.

	•	Refactored the UniversalDocumentLoader to use MIME type detection exclusively, leveraging the existing MimeTypeBasedParser to dynamically select parsers based on MIME type rather than file extensions, in line with the design in langchain-community/libs/community/langchain_community/document_loaders/parsers/generic.py.
	•	Removed hard-coded extension based logic to enhance extensibility and allow better support for multiple document types as they grow.
	•	Addressed the concern about the large .rtf file by replacing it with a minimal, representative sample, reducing test resource size and improving test efficiency.
	•	Updated relevant imports and parser registrations to align with this MIME type driven approach.

I believe these changes adhering to the existing framework and make the loader more flexible and maintainable. Please let me know if further adjustments are needed.

aiml-learning-design commented on libs/community/langchain_community/universal_document_loader/custom_parsers/parser_csv.py:
```diff
@@ -0,0 +1,12 @@
+    def lazy_parse(self, blob: Blob) -> List[Document]:
+        """"""
+        df = pd.read_csv(blob.path)
+        return [Document(page_content=df.to_csv(index=False), metadata={"source": str(blob.path)})]
```
> The `lazy_parse` method is missing a proper docstring. Consider adding a Google-style docstring that describes the method's purpose and parameters:
> 
> ```python
> def lazy_parse(self, blob: Blob) -> List[Document]:
>     """Parse CSV file into a Document.
>     
>     Args:
>         blob: Blob object containing the CSV file path
>         
>     Returns:
>         List[Document]: A list containing a single Document with the CSV content
>     """
> ```
> 
> This will improve code readability and help maintain documentation standards across the codebase.
> 
> _Spotted by [Diamond](https://app.graphite.dev/diamond/?org=langchain-ai&ref=ai-review-comment) (based on [custom rules](https://app.graphite.dev/diamond/?org=langchain-ai&ref=ai-review-comment&view=settings))_Is this helpful? React 👍 or 👎 to let us know.

DocString has been updated as suggested

eyurtsev reviewed (changes requested):
Hi @aiml-learning-design! I don't think we're likely to merge this PR. I don't want an entrypoint to document loading that assumes files are read from disk. The BlobParser abstraction that exists already implements the correct design.

It's poorly documented right now: https://python.langchain.com/docs/how_to/document_loader_custom/#working-with-files
